### PR TITLE
[oneDNN] Improving Graph Rewrite Performance

### PR DESCRIPTION
**Pull request by mahmoud-abuzaina**
This PR improves the latency of graph rewrite by avoiding doing kernel registry lookup more than once per kernel. Similar fix was to done before to eager mode, so the refactoring is done to manage and maintain that kernel registry info in one place (mkl_graph_util.h)

### CODE REVIEW

**Comment by gbaned** [comment]
@penpornk  Can you please review this PR ? Thanks!

File: tensorflow/core/graph/mkl_graph_util.h [hunk 1]
```diff
@@ -148,6 +148,10 @@ static const char* const kMklEagerOpPrefix = "_MklEager";
 // _MklEager prefix.
 static const char* const kMklNativeOpPrefix = "_MklNative";
 
+// Map used to store kernel registration status
+thread_local static auto* registered_kernels_map =
+    new absl::flat_hash_map<string, bool>();
```
Comment (penpornk) [inline tensorflow/core/graph/mkl_graph_util.h:153]: Does this need to be a global static variable? We try to only have static variable when they are [trivially destructible](https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables). Can this map be part of some class?

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you very much for the fix and I'm so sorry for my delay!

File: tensorflow/core/graph/mkl_graph_util.h [hunk 2]
```diff
@@ -148,6 +148,10 @@ static const char* const kMklEagerOpPrefix = "_MklEager";
 // _MklEager prefix.
 static const char* const kMklNativeOpPrefix = "_MklNative";
 
+// Map used to store kernel registration status
+thread_local static std::unordered_map<std::string, bool>
```
[the remark this answers is not in the document]
Comment (penpornk) [reply tensorflow/core/graph/mkl_graph_util.h]: Please ignore this comment. I left this comment when I was first looking at the PR a month ago and didn't have a chance to finish it. It's outdated now.

File: tensorflow/core/graph/mkl_graph_util.h [hunk 3]
```diff
@@ -196,105 +200,85 @@ static inline void BF16UnsupportedWarning() {
...
+static inline bool IsMklOp(const string& op_name, DataType T,
+                           bool is_native_op) {
+  string label = is_native_op ? kMklNameChangeOpLabelPattern
+                              : kMklLayoutDependentOpLabelPattern;
+  string registered_kernels_key = op_name + label + std::to_string(T);
+  auto kernel_element = registered_kernels_map.find(registered_kernels_key);
+  bool kernel_registered = false;
+
+  if (kernel_element == registered_kernels_map.end()) {
+    string kernel = KernelsRegisteredForOp(op_name);
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/graph/mkl_graph_util.h]: Done.

File: tensorflow/core/graph/mkl_graph_util.h [hunk 4]
```diff
@@ -196,105 +200,85 @@ static inline void BF16UnsupportedWarning() {
...
+    if (is_native_op &&
+        kernel.find(kMklQuantizedOpLabelPattern) != string::npos) {
+      // Restrict quantized ops to QUINT8, QINT8 and DT_QINT32
+      kernel_registered = (T == DT_QUINT8 || T == DT_QINT8 || T == DT_QINT32);
+    }
+
+    // Now we just construct a search string to match what we are looking for.
+    string search_string = label;
+    search_string += string(";") + string(" T in [");
+    search_string += DataType_Name(T) + string("]");
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/graph/mkl_graph_util.h]: Done.

File: tensorflow/core/graph/mkl_graph_util.h (hunk 1 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/graph/mkl_graph_util.h]: yes it can be in `IsMklOp()`. I made that change.

**Comment by mahmoud-abuzaina** [comment]
Thank you for reviewing the PR. I have addressed your comments.

**Review by penpornk — APPROVED** [review]
Thank you for the changes!

**Pull request merged**